picasso/src/main/java/com/squareup/picasso3/BitmapHunter.java

```diff
@@ -209,52 +212,70 @@ Bitmap hunt() throws IOException {
+    catch (InterruptedException ie) {
+      throw new IOException(ie);
     }
 
-    if (bitmap != null) {
-      if (picasso.loggingEnabled) {
-        log(OWNER_HUNTER, VERB_DECODED, data.logId());
-      }
-      stats.dispatchBitmapDecoded(bitmap);
-      if (data.needsTransformation() || exifOrientation != 0) {
-        synchronized (DECODE_LOCK) {
-          if (data.needsMatrixTransform() || exifOrientation != 0) {
-            bitmap = transformResult(data, bitmap, exifOrientation);
-            if (picasso.loggingEnabled) {
-              log(OWNER_HUNTER, VERB_TRANSFORMED, data.logId());
-            }
-          }
-          if (data.hasCustomTransformations()) {
-            bitmap = applyCustomTransformations(data.transformations, bitmap);
-            if (picasso.loggingEnabled) {
-              log(OWNER_HUNTER, VERB_TRANSFORMED, data.logId(), "from custom transformations");
+    requestHandler.load(data, networkPolicy, new Callback() {
+      @Override public void onSuccess(RequestHandler.Result result) {
+        Bitmap bitmap = null;
+        if (result != null) {
+          loadedFrom = result.getLoadedFrom();
+          exifOrientation = result.getExifOrientation();
+          bitmap = result.getBitmap();
+
+          // If there was no Bitmap then we need to decode it from the stream.
+          if (bitmap == null) {
+            Source source = result.getSource();
+            try {
+              bitmap = decodeStream(source, data);
+            } finally {
+              try {
+                //noinspection ConstantConditions If bitmap is null then source is guranteed non-null.
+                source.close();
+              } catch (IOException ignored) {
+              }
             }
           }
         }
+
         if (bitmap != null) {
-          stats.dispatchBitmapTransformed(bitmap);
+          if (picasso.loggingEnabled) {
+            log(OWNER_HUNTER, VERB_DECODED, data.logId());
+          }
+          stats.dispatchBitmapDecoded(bitmap);
+          if (data.needsTransformation() || exifOrientation != 0) {
+            synchronized (DECODE_LOCK) {
+              if (data.needsMatrixTransform() || exifOrientation != 0) {
+                bitmap = transformResult(data, bitmap, exifOrientation);
+                if (picasso.loggingEnabled) {
+                  log(OWNER_HUNTER, VERB_TRANSFORMED, data.logId());
+                }
+              }
+              if (data.hasCustomTransformations()) {
+                bitmap = applyCustomTransformations(data.transformations, bitmap);
+                if (picasso.loggingEnabled) {
+                  log(OWNER_HUNTER, VERB_TRANSFORMED, data.logId(), "from custom transformations");
+                }
+              }
+            }
+            if (bitmap != null) {
+              stats.dispatchBitmapTransformed(bitmap);
+            }
+          }
         }
       }
-    }
+
+      @Override public void onError(Throwable t) {
+        throw new IOException(t);
```

Comment:
what's going on here?
this looks like it's meant to throw in BitmapHunter.load. is this supposed to be in an `AtomicReference<Exception>`?

Reply:
good point.  Yes, it should throw in BitmapHunter.  Thoughts on wrapping in a `RuntimeException `which would get propagated back up to `BitmapHunter.load()`?

Reply:
Updated to use an AtomicReference<Throwable>.